File: src/dotnet/PloyBus.Tests/Transport/InMemory/TestEnvironment.cs
using Microsoft.Extensions.Logging;
using PolyBus.Transport.Transactions.Messages.Handlers.Serializers;

namespace PolyBus.Transport.InMemory;

class TestEnvironment
{
    public ILoggerFactory LoggerFactory { get; } = Microsoft.Extensions.Logging.LoggerFactory.Create(builder =>
    {
        builder.AddProvider(new TestContextLoggerProvider());
    });
    public InMemoryMessageBroker InMemoryMessageBroker { get; } = new();
    public TestEndpoint Alpha { get; } = new();
    public TestEndpoint Beta { get; } = new();
    public async Task Setup()
    {
        await SetupEndpoint(Alpha, "alpha");
        await SetupEndpoint(Beta, "beta");
    }

    async Task SetupEndpoint(TestEndpoint testEndpoint, string name)
    {
        var jsonHandlers = new JsonHandlers();

        // add handlers for incoming messages
        testEndpoint.Builder.IncomingPipeline.Add(jsonHandlers.Deserializer);
        testEndpoint.Builder.IncomingPipeline.Add(testEndpoint.Handler);

        // add messages
        testEndpoint.Builder.Messages.Add(typeof(AlphaCommand));
        testEndpoint.Builder.Messages.Add(typeof(AlphaEvent));
        testEndpoint.Builder.Name = name;

        // add handlers for outgoing messages
        testEndpoint.Builder.OutgoingPipeline.Add(jsonHandlers.Serializer);

        // configure InMemory transport
        testEndpoint.Builder.TransportFactory = InMemoryMessageBroker.AddEndpoint;
        InMemoryMessageBroker.Log = LoggerFactory.CreateLogger<InMemoryMessageBroker>();

        // create the bus instance
        testEndpoint.Bus = await testEndpoint.Builder.Build();
    }

    public async Task Start()
    {
        await Alpha.Bus.Start();
        await Beta.Bus.Start();
    }

    public async Task Stop()
    {
        await Alpha.Bus.Stop();
        await Beta.Bus.Stop();
    }
}

File: src/kotlin/src/test/kotlin/polybus/transport/inmemory/TestEnvironment.kt
package polybus.transport.inmemory

import polybus.transport.transactions.messages.handlers.serializers.JsonHandlers
import java.util.logging.Logger

class TestEnvironment {
    val inMemoryMessageBroker: InMemoryMessageBroker = InMemoryMessageBroker()
    val alpha: TestEndpoint = TestEndpoint()
    val beta: TestEndpoint = TestEndpoint()

    suspend fun setup() {
        setupEndpoint(alpha, "alpha")
        setupEndpoint(beta, "beta")
    }

    private suspend fun setupEndpoint(testEndpoint: TestEndpoint, name: String) {
        val jsonHandlers = JsonHandlers()

        // add handlers for incoming messages
        testEndpoint.builder.incomingPipeline.add(jsonHandlers::deserializer)
        testEndpoint.builder.incomingPipeline.add(testEndpoint::handler)

        // add messages
        testEndpoint.builder.messages.add(AlphaCommand::class.java)
        testEndpoint.builder.messages.add(AlphaEvent::class.java)
        testEndpoint.builder.name = name

        // add handlers for outgoing messages
        testEndpoint.builder.outgoingPipeline.add(jsonHandlers::serializer)

        // configure InMemory transport
        testEndpoint.builder.transportFactory = inMemoryMessageBroker::addEndpoint
        inMemoryMessageBroker.log = Logger.getLogger(InMemoryMessageBroker::class.java.name)

        // create the bus instance
        testEndpoint.bus = testEndpoint.builder.build()
    }

    suspend fun start() {
        alpha.bus.start()
        beta.bus.start()
    }

    suspend fun stop() {
        alpha.bus.stop()
        beta.bus.stop()
    }
}
